File: solutions/python/robot-name/1/robot_name.py
import random
import string

class Robot:
    used_names = []

    def __init__(self):
        self._name = self._generate_unique_name()

    @property
    def name(self):
        return self._name

    def reset(self):
        self._name = self._generate_unique_name()

    def _generate_name(self):
        letter_one = random.choice(string.ascii_uppercase)
        letter_two = random.choice(string.ascii_uppercase)
        number_code = str(random.randint(0, 999)).zfill(3)
        return f"{letter_one}{letter_two}{number_code}"

    def _generate_unique_name(self):
        while True:
            new_name = self._generate_name()
            if new_name not in Robot.used_names:
                Robot.used_names.append(new_name)
                return new_name

    @classmethod
    def release_names(cls):
        cls.used_names = []

File: solutions/cpp/robot-name/1/robot_name.cpp
#include "robot_name.h"

#include <iomanip>
#include <random>

namespace robot_name {
    std::set<std::string> robot::robot_names{};
    std::string robot::generate_random_name() {
        static std::random_device rd;
        static std::mt19937 gen(rd());

        std::uniform_int_distribution<> letter_dist(0, 25);
        std::uniform_int_distribution<> num_dist(0, 999);

        const char letter1 = static_cast<char>('A' + letter_dist(gen));
        const char letter2 = static_cast<char>('A' + letter_dist(gen));
        const int num = num_dist(gen);

        std::ostringstream oss;
        oss << letter1
                << letter2
                << std::setw(3)
                << std::setfill('0')
                << num;
        return oss.str();
    }

    std::string robot::generate_unique_name() {
        std::string newName;
        do {
            newName = generate_random_name();
        } while (robot_names.count(newName) != 0);
        robot_names.insert(newName);
        return newName;
    }

    robot::robot() {
        robot_name = generate_unique_name();
    }

    std::string robot::name() const {
        return robot_name;
    }

    void robot::reset() {
        robot_name = generate_unique_name();
    }
} // namespace robot_name
